sql-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/WindowedBlockIterator.scala

```diff
@@ -0,0 +1,174 @@
+// as long as they can provide a size
+trait BlockWithSize {
+  /**
+   * Abstract method to return the size in bytes of this block
+   * @return Long - size in bytes
+   */
+  def size: Long
+}
+
+/**
+ * Specifies a start and end range of bytes for a block.
+ * @param block - a BlockWithSize instance
+ * @param rangeStart - byte offset for the start of the range
+ * @param rangeEnd - byte offset for the end of the range
+ * @tparam T - the specific type of `BlockWithSize`
+ */
+case class BlockRange[T <: BlockWithSize](
+    block: T, rangeStart: Long, rangeEnd: Long) {
+  require(rangeStart <= rangeEnd,
+    s"Instantiated a BlockRange with invalid boundaries: $rangeStart to $rangeEnd")
+
+  /**
+   * Returns the size of this range in bytes
+   * @return - Long - size in bytes
+   */
+  def rangeSize(): Long = rangeEnd - rangeStart + 1
+
+  def isComplete(): Boolean = rangeEnd == block.size - 1
+}
+
+/**
+ * Given a set of blocks, this iterator returns BlockRanges
+ * of such blocks that fit `windowSize`. The ranges are just logical
+ * chunks of the blocks, so this class performs no memory management or copying.
+ *
+ * If a block is too large for the window, the block will be
+ * returned in `next()` until the full block can be covered.
+ *
+ * For example, given a block that is 4 window-sizes in length:
+ * block = [sb1, sb2, sb3, sb4]
+ *
+ * The window will return on `next()` four "sub-blocks", governed by `windowSize`:
+ * window.next() // sb1
+ * window.next() // sb2
+ * window.next() // sb3
+ * window.next() // sb4
+ *
+ * If blocks are smaller than the `windowSize`, they will be packed:
+ * block1 = [b1]
+ * block2 = [b2]
+ * window.next() // [b1, b2]
+ *
+ * A mix of both scenarios above is possible:
+ * block1 = [sb11, sb12, sb13] // where sb13 is smaller than window length
+ * block2 = [b2]
+ *
+ * window.next() // sb11
+ * window.next() // sb12
+ * window.next() // [sb13, b2]
+ *
+ * @param blocks - sequence of blocks to manage
+ * @param windowSize - the size (in bytes) that block ranges should fit
+ * @tparam T - the specific type of `BlockWithSize`
+ * @note this class does not own `transferBlocks`
+ * @note this class is not thread safe
+ */
+class WindowedBlockIterator[T <: BlockWithSize](blocks: Seq[T], windowSize: Long)
+    extends Iterator[Seq[BlockRange[T]]] {
+
+  require(windowSize > 0, s"Invalid window size specified $windowSize")
+
+  case class BlockWindow(start: Long, size: Long) {
+    val end = start + size - 1
+    def move(): BlockWindow = {
+      BlockWindow(start + size, size)
+    }
+  }
+
+  // start the window at byte 0
+  private[this] var window = BlockWindow(0, windowSize)
+  private[this] var done = false
+
+  // helper class that captures the start/end byte offset
+  // for `block` on creation
+  case class BlockWithOffset[T <: BlockWithSize](
+      block: T, startOffset: Long, endOffset: Long)
+
+  private var lastOffset = 0L
+  private[this] val blocksWithOffsets = blocks.map { block =>
```

Comment:
lastOffset doesn't need to be a class member since it's only needed to initialize `blocksWithOffsets`.  Can update to something like this:
```scala
private[this] val blocksWithOffsets = {
  var lastOffset = 0L
  blocks.map { block =>
   ...
  }
}
```

Reply:
done.